File: 567-Permutation in String.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        dict1 = [0]*26
        dict2 = [0]*26
        
        for i in range(len(s1)):
            dict1[ord(s1[i]) - ord('a')] += 1
        N = len(s1)
        for j in range(len(s2)):
            dict2[ord(s2[j]) - ord('a')] += 1
            
            if j >= N:
                dict2[ord(s2[j-N]) - ord('a')] -= 1
            
            if dict1 == dict2:
                return True
        return False

File: C++/567-Permutation in String.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int cnt = 0, N = s1.size(), left = 0;
        int ch[26] = {0};
        for (char c:s1) ch[c-'a']++;

        for (int i=0; i<s2.size(); i++) {
            int num = s2[i]-'a';
            ch[num]--;
            if (ch[num] >= 0) cnt++;
            if (ch[num] < 0) {
                while (left <= i && i-left+1 != cnt) {
                    ch[s2[left]-'a']++;
                    if (ch[s2[left]-'a'] > 0) cnt--;
                    left++;
                }
            }
            if (cnt == N) return true;
        }

        return false;
    }
};